library/rsa.c

```diff
@@ -109,6 +109,10 @@ int mbedtls_rsa_parse_key(mbedtls_rsa_context *rsa, const unsigned char *key, si
 
     end = p + len;
 
+    if (end > (key + keylen)) {
```

Comment:
This check never fires: at this point, the comparison is between the end of the SEQUENCE content and the end of the overall buffer, and a success from `mbedtls_asn1_get_tag` guarantees that they're in the desired order. I think the fix is to put this exact check just above `end = p + len`.

Same with `parse_pubkey`.

Reply:
> success from mbedtls_asn1_get_tag guarantees that they're in the desired order

Ops, that's a detail that I missed before. Thanks for the hint!

Reply:
> I think the fix is to put this exact check just above end = p + len.

Mmmm, but (if I'm not wrong) doing so would result in a never-firing-check again. `end` is set few lines above as `end = p + keylen;` (where `p = (unsigned char *) key;` at that point) and then `mbedtls_asn1_get_tag()` does not update it, so checking `if (end > (key + keylen))` never fires. Am I wrong?

On the other hand if I replace `if (end > (key + keylen))` with `if (end != (key + keylen))` there are other test suites that start failing. For example `test_suite_pkparse` fails at:
```
Parse RSA Key #2 (Correct password) ............................... FAILED
Parse RSA Key #4 (DES Encrypted) .................................. FAILED
Parse RSA Key #5 (3DES Encrypted) ................................. FAILED
Parse RSA Key #6 (AES-128 Encrypted) .............................. FAILED
Parse RSA Key #7 (AES-192 Encrypted) .............................. FAILED
Parse RSA Key #8 (AES-256 Encrypted) .............................. FAILED
Parse RSA Key #9 (2048-bit, DES Encrypted) ........................ FAILED
Parse RSA Key #10 (2048-bit, 3DES Encrypted) ...................... FAILED
Parse RSA Key #11 (2048-bit, AES-128 Encrypted) ................... FAILED
Parse RSA Key #12 (2048-bit, AES-192 Encrypted) ................... FAILED
Parse RSA Key #13 (2048-bit, AES-256 Encrypted) ................... FAILED
Parse RSA Key #14 (4096-bit, DES Encrypted) ....................... FAILED
Parse RSA Key #15 (4096-bit, 3DES Encrypted) ...................... FAILED
Parse RSA Key #16 (4096-bit, AES-128 Encrypted) ................... FAILED
Parse RSA Key #17 (4096-bit, AES-192 Encrypted) ................... FAILED
Parse RSA Key #18 (4096-bit, AES-256 Encrypted) ................... FAILED
```

I will have a look at this of course, but I'm not sure that having the extra data is a real bad thing (see my [previous comment](https://github.com/Mbed-TLS/mbedtls/pull/8740#discussion_r1480917477)). Wdyt?

Reply:
There might be some history of sloppiness in pk parsing of encrypted keys, because the decryption code (`mbedtls_pkcs5_pbes2` and `mbedtls_pkcs12_pbe`) used to not report the length of the plaintext. But nowadays pkparse calls `mbedtls_pkcs5_pbes2_ext` and `mbedtls_pkcs12_pbe_ext`, which do report the length of the plaintext. So I'd expect this to be easy to fix.

Reply:
The failing cases I reported above are not related to pkcs5/12. For example in
`Parse RSA Key #2 (Correct password)`
the flow is:
`mbedtls_pk_parse_key -> mbedtls_pem_read_buffer -> pem_aes_decrypt`

Here `len` is not updated at the end of `mbedtls_pem_read_buffer` compared to what it was allocated from the  [first call](https://github.com/Mbed-TLS/mbedtls/blob/1d7bc1ecdf04db9b137931442fd46fe278813f2b/library/pem.c#L386) to `mbedtls_base64_decode()` on the encrypted buffer. As a consequence the following call to `mbedtls_rsa_parse_key` fails.
The only way to prevent this would be to analyze the decrypted buffer to determine how long it is based on the SEQUENCE data, but that's part of the RSA parser in my opinion. Wdyt?

Reply:
Oh, I'd forgotten we also had a decryption method in `pem.c`. It's a similar construction as PKCS5 and PKCS12: CBC where the old decryption code doesn't strip away the padding. For PKCS5 and PKCS12, we added `_ext` functions that do strip the padding. So I guess we should do the same thing in `pem.c`: remove PKCS7 padding after CBC decryption.

Since `pem.c` can be built without `MBEDTLS_CIPHER_C`, we can't use the `cipher.c` code (`get_pkcs_padding`) to strip the padding. On the other hand, since we're parsing the content, we don't have to worry about a padding oracle: there's a decryption oracle via the content anyway. So the padding check doesn't need to be constant-time.

Reply:
To move on with this pull request which is already pretty long, I propose to merge it as is, and fix this in a follow-up.